### PR TITLE
Ddp 5045 sendgrid events

### PR DESCRIPTION
**Pull request by evgeniipr**
## Context

Implemented REST endpoint: POST ../pepper/v1/sendgrid-event which persists the SendGrid events passed from SendGrid.
In order to use this feature it needs to configure webhook in SendGrid in this URL: https://app.sendgrid.com/settings/mail_settings
The backend logic organized the following way:
**SendGridEventRoute** delegates the logic of parsing log events and persisting them to **SendGridEventService**.
SendGrid event could be signed and signature verification is checked by **SendGridEventVerificationFilter**.
It's created a Liquibase script to create a BD table: **sendgrid_event** where incoming events will be saved.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [x] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [x] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [x] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [x] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [x] I have written automated positive tests
- [x] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [x] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ ] These changes require no special release procedures--just code!
- [x] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/src/main/resources/db-changes/schema/DDP-5045-sendgrid-event.xml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                <constraints nullable="false"/>
+            </column>
+            <column name="url" type="varchar(1000)">
+                <constraints nullable="true"/>
+            </column>
+            <column name="ip" type="varchar(100)">
+                <constraints nullable="true"/>
+            </column>
+            <column name="sg_event_id" type="varchar(100)">
+                <constraints nullable="false" unique="true" uniqueConstraintName="sendgrid_event_sg_event_id_uk" />
```
Comment (yufengwng) [inline pepper-apis/src/main/resources/db-changes/schema/DDP-5045-sendgrid-event.xml:30]: Is `sg_event_id` unique across all of SendGrid? What if we have 2 different SendGrid accounts sending events to us?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/SendGridEventRoute.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import spark.Route;
+
+public class SendGridEventRoute implements Route {
+
+    private static final Logger LOG = getLogger(SendGridEventRoute.class);
+
+    private final SendGridEventService sendGridEventService;
+
+    public SendGridEventRoute() {
+        this.sendGridEventService = new SendGridEventService();
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/route/SendGridEventRoute.java]: Can we pass this in as a dependency?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/SendGridEventRoute.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        }
+        response.status(SC_OK);
+        return "";
+    }
+
+    private void persistLogEvent(SendGridEvent[] sendGridEvents) {
+        try {
+            TransactionWrapper.useTxn(handle -> sendGridEventService.persistLogEvents(handle, sendGridEvents));
+        } catch (Exception e) {
+            haltError(SC_INTERNAL_SERVER_ERROR, DATA_PERSIST_ERROR, e.getMessage());
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/route/SendGridEventRoute.java]: The exception message might be too low-level to return as a message to the API caller. Maybe consider returning a more generic message string, and then logging the exception with `LOG.error()` so our error reporting pick up on it.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/json/sendgrid/SendGridEventTypeAdapter.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import com.google.gson.TypeAdapter;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonToken;
+import com.google.gson.stream.JsonWriter;
+
+/**
+ * Convert UNIX timestamp to Java Instant
+ */
+public class SendGridEventTypeAdapter extends TypeAdapter<Instant> {
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/json/sendgrid/SendGridEventTypeAdapter.java]: We have a few of these time-related adapters in the `ddp.transformers` package. Can you try putting this in there and generalizing it a bit so we can reuse it? E.g. implement `write()` method and rename to something like `InstantToEpochSecondsAdapter`.

File: pepper-apis/src/test/resources/sendgrid-event-testdata/sendgrid-event-testdata.json [hunk 5]
```diff
@@ -0,0 +1,138 @@
+[
+  {
+      "email":"example1@test.com",
```
Comment (yufengwng) [inline pepper-apis/src/test/resources/sendgrid-event-testdata/sendgrid-event-testdata.json]: I recommend using `@datadonationplatform.org` for dummy emails.

**Review by yufengwng — COMMENTED** [review]
Overall looks good.

File: pepper-apis/config/application.conf.ctmpl [hunk 6]
```diff
@@ -32,6 +32,11 @@
         "dsmClientId": "{{$auth0.dsmClientId}}",
         "dsmClientSecret": "{{$auth0.dsmClientSecret}}"
     },
+    "sendGridEvents" : {
+	{{if $conf.Data.sendGridEvents.verificationKey}}
+		"verificationKey" : "{{$conf.Data.sendGridEvents.verificationKey}}"
```
Comment (yufengwng) [inline pepper-apis/config/application.conf.ctmpl]: We have a `sendgrid` config object below. Will it makes sense to group this new property in there? E.g. as a new `sendgrid.eventsVerificationKey` property.

File: pepper-apis/src/main/resources/db-changes/schema/DDP-5045-sendgrid-event.xml (hunk 1 above)
Comment (evgeniipr) [reply pepper-apis/src/main/resources/db-changes/schema/DDP-5045-sendgrid-event.xml:30]: As I understand it should be unique across all of SendGrid. Example of such ID: "zmzJhfJgAfUSOW80yEbPyw==".

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/SendGridEventRoute.java (hunk 2 above)
Comment (evgeniipr) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/route/SendGridEventRoute.java]: Done

File: pepper-apis/src/main/resources/db-changes/schema/DDP-5045-sendgrid-event.xml (hunk 1 above)
Comment (zyme) [reply pepper-apis/src/main/resources/db-changes/schema/DDP-5045-sendgrid-event.xml:30]: Is there a way to associate each event with the `umbrella_study`?  I'm wondering if we passed more data into our API calls when we send emails, could we get that data back in some way in the logs?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/SendGridEventRoute.java (hunk 3 above)
Comment (evgeniipr) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/route/SendGridEventRoute.java]: Fixed

File: pepper-apis/src/main/java/org/broadinstitute/ddp/json/sendgrid/SendGridEventTypeAdapter.java (hunk 4 above)
Comment (evgeniipr) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/json/sendgrid/SendGridEventTypeAdapter.java]: Refactored. Created UnixTimestampToInstantAdapter.java

File: pepper-apis/src/test/resources/sendgrid-event-testdata/sendgrid-event-testdata.json (hunk 5 above)
Comment (evgeniipr) [reply pepper-apis/src/test/resources/sendgrid-event-testdata/sendgrid-event-testdata.json]: Fixed

File: pepper-apis/config/application.conf.ctmpl (hunk 6 above)
Comment (evgeniipr) [reply pepper-apis/config/application.conf.ctmpl]: I see in config that there are several params which also not belong to "sendgrid" section:
    "sendgridToken": "sendgridToken",
    "sendgridTemplates": {
But in general I agree that it makes sense.
But seems it needs to do the same with auth0LogEvents.token - to move it under auth0 section?

File: pepper-apis/src/main/resources/db-changes/schema/DDP-5045-sendgrid-event.xml (hunk 1 above)
Comment (evgeniipr) [reply pepper-apis/src/main/resources/db-changes/schema/DDP-5045-sendgrid-event.xml:30]: @zyme yes, we could define a "category", and then send it back. But needs to experiment. I will check it tomorrow

File: pepper-apis/config/application.conf.ctmpl (hunk 6 above)
Comment (evgeniipr) [reply pepper-apis/config/application.conf.ctmpl]: Moved the parameter into 'sendgrid' section and renamed as suggested.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/transformers/UnixTimestampToInstantAdapter.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        }
+    }
+
+    @Override
+    public Instant read(JsonReader in) throws IOException {
+        if (in.peek() == JsonToken.NULL) {
+            in.nextNull();
+            return null;
+        }
+        return Instant.ofEpochSecond(in.nextInt());
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/transformers/UnixTimestampToInstantAdapter.java]: Will it be safer to use `in.nextLong()` here instead?

File: pepper-apis/src/test/java/org/broadinstitute/ddp/service/SendGridEventServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Test
+    public void testAuth0LogEventsParse() throws FileNotFoundException {
+        var sendGridEvents = sendGridEventService.parseSendGridEvents(readJsonArrayFromFile(SENDGRID_EVENT_TESTDATA).toString());
+
+        assertEquals(12, sendGridEvents.length);
+
+        var sendGridEvent = sendGridEvents[0];
+        assertEquals("example1@datadonationplatform.org", sendGridEvent.getEmail());
+        assertEquals("2017-12-15T00:59:29Z", String.valueOf(sendGridEvent.getTimestamp()));
```
Comment (yufengwng) [inline pepper-apis/src/test/java/org/broadinstitute/ddp/service/SendGridEventServiceTest.java]: Simpler way is to call `toString()` on the `Instant` objects.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/SendGridEventRoute.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    private void checkBody(Request request) {
+        if (StringUtils.isBlank(request.body())) {
+            haltError(SC_BAD_REQUEST, MISSING_BODY, "Body not specified", null);
+        }
+    }
+
+    private void haltError(int status, String code, String msg, Exception e) {
+        LOG.warn(msg + (e != null ? ' ' + e.getMessage() : ""));
```
Comment (yufengwng) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/route/SendGridEventRoute.java]: If `e` is not null, we should log the whole exception object too. I think you can just pass `e` as argument to `LOG.warn()` call, for example:
```
LOG.warn(msg + (...), e);
```

File: pepper-apis/src/main/java/org/broadinstitute/ddp/transformers/UnixTimestampToInstantAdapter.java (hunk 7 above)
Comment (evgeniipr) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/transformers/UnixTimestampToInstantAdapter.java]: Yes, agree. Fixed

File: pepper-apis/src/test/java/org/broadinstitute/ddp/service/SendGridEventServiceTest.java (hunk 8 above)
Comment (evgeniipr) [reply pepper-apis/src/test/java/org/broadinstitute/ddp/service/SendGridEventServiceTest.java]: Fixed

File: pepper-apis/src/main/java/org/broadinstitute/ddp/route/SendGridEventRoute.java (hunk 9 above)
Comment (evgeniipr) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/route/SendGridEventRoute.java]: Agree. Fixed

Comment (evgeniipr) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/route/SendGridEventRoute.java]: Fixed

**Review by yufengwng — DISMISSED** [review]
LGTM

**Review by yufengwng — APPROVED** [review]

**Pull request merged**